### PR TITLE
Added typesafe config usage and refactored server state machine

### PR DESCRIPTION
**Pull request by shnapz**

### CODE REVIEW

File: example/src/main/scala/pulse/services/example/CliSettings.scala [hunk 1]
```diff
@@ -2,19 +2,19 @@ package pulse.services.example
...
-    parser.parse(args, Config()) match {
+    parser.parse(args, CliSettings()) match {
       case Some(config) => config
       case None => throw new InvalidCmdArgsException("Invalid arguments passed")
     }
   }
 }
 
-case class Config(useTaskApi: Boolean = false)
+case class CliSettings(useTaskApi: Boolean = false)
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/CliSettings.scala]: why clisettings?

File: example/src/main/scala/pulse/services/example/Settings.scala [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.typesafe.config.Config
+
+
+/**
+  * Created by Andrew on 19.03.2017.
+  */
+class Settings(config: Config, cliSettings: CliSettings) {
+
+  val threadPoolMaxCount: Int = config.getInt("threadpool.maxCount")
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/Settings.scala]: that's bad, it's not composable

File: core/src/main/scala/pulse/services/core/package.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package pulse.services
+
+import cats.MonadError
+import com.twitter.util.{Awaitable, Future}
+
+/**
+  * Created by Andrew on 19.03.2017.
+  */
+package object core {
+  implicit val future = new MonadError[Future, Throwable] {
```
Comment (theocat) [inline core/src/main/scala/pulse/services/core/package.scala]: is it related to the typesafe?

Comment (shnapz) [inline core/src/main/scala/pulse/services/core/package.scala]: Oops, it's part of state machine refactoring

File: example/src/main/scala/pulse/services/example/Settings.scala (hunk 2 above)
Comment (shnapz) [inline example/src/main/scala/pulse/services/example/Settings.scala]: What exactly do you mean? It is a pattern from typesafe documentation

File: core/src/main/scala/pulse/services/core/ServerApp.scala [hunk 4]
```diff
@@ -2,85 +2,62 @@ package pulse.services
...
   def server(args: List[String]): ListeningServer
 
   def main(): Unit = {
-    getOrLogError(stateMachine.transitTo(Starting)) { startingState =>
-
-      Await.ready(startingState)
-      serverInstance = server(args.toList)
-      Await.ready(serverInstance)
-    }
+    Await.ready(Await.result(startServer()))
```
Comment (theocat) [inline core/src/main/scala/pulse/services/core/ServerApp.scala]: Await.ready Await.result?

File: core/src/main/scala/pulse/services/core/StateMachine.scala [hunk 5]
```diff
@@ -1,30 +1,53 @@
...
-      case _ => transitTo(newState)
-    }
+
+trait StateMachine[Step, State] {
+
+  case class Transition(currentEvent: Step, nextEvent: Step, state: State)
+
+  type TransitionFunction = PartialFunction[Transition, Future[State]]
+
+  def initialState: Step <> State
```
Comment (theocat) [inline core/src/main/scala/pulse/services/core/StateMachine.scala]: why def?

File: core/src/main/scala/pulse/services/core/ServerApp.scala (hunk 4 above)
Comment (shnapz) [inline core/src/main/scala/pulse/services/core/ServerApp.scala]: We have a Future and listeningServer inside. It is awaitable

File: core/src/main/scala/pulse/services/core/StateMachine.scala (hunk 5 above)
Comment (shnapz) [inline core/src/main/scala/pulse/services/core/StateMachine.scala]: Because val will be null from the trait initialization code, where I use this value

Comment (shnapz) [inline core/src/main/scala/pulse/services/core/StateMachine.scala]: Value from def initializes vars on 48, 50

**Pull request merged**